File: leetcode/Easy/13. Roman to Integer.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {

public:
    int romanToInt(string s) {
        int sum = 0, len = s.length();
        map<char, int> mp;
        
        mp['I'] = 1, mp['V'] = 5, mp['X'] = 10;
        mp['L'] = 50, mp['C'] = 100, mp['D'] = 500, mp['M'] = 1000;
        
        for (int i = len - 1; i >= 0; i--) {
            char ch = s[i];
            sum += mp[ch];
            if (i < len - 1) {
                if (ch == 'I') {
                    if (s[i + 1] == 'V' || s[i + 1] == 'X') {
                        sum -= 2;
                    }
                } else if (ch == 'X') {
                    if (s[i + 1] == 'L' || s[i + 1] == 'C') {
                        sum -= 20;
                    }
                } else if (ch == 'C') {
                    if (s[i + 1] == 'D' || s[i + 1] == 'M') {
                        sum -= 200;
                    }
                }
            }
        }
        return sum;
    }
};

File: leetcode/Easy/13. Roman to Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        dic = { 'I':1, 'V':5, 'X':10, 'L':50, 'C':100, 'D':500, 'M':1000 }
        ret = 0
        
        for idx in range(len(s) - 1):
            if dic[s[idx]] < dic[s[idx + 1]]:
                ret -= dic[s[idx]]
            else:
                ret += dic[s[idx]]
        
        ret += dic[s[-1]]
        return ret